internal/admission/validator_test.go

```diff
@@ -288,4 +295,218 @@ func TestKongHTTPValidator_ValidateClusterPlugin(t *testing.T) {
+		secrets                        []*corev1.Secret
+		consumers                      []*configurationv1.KongConsumer
+		modifyBasicConsumer            func(c *configurationv1.KongConsumer)
+		operation                      admissionv1.Operation
+		consumerAlreadyExistsInGateway bool
+		expectError                    bool
+		expectOK                       bool
+		expectedMessage                string
+	}{
+		{
+			name: "consumer refers a non-existent secret",
+			modifyBasicConsumer: func(c *configurationv1.KongConsumer) {
+				c.Credentials = []string{"non-existing-secret"}
+			},
+			operation:       admissionv1.Create,
+			expectOK:        false,
+			expectError:     false,
+			expectedMessage: "consumer referenced non-existent credentials secret: default/non-existing-secret",
+		},
+		{
+			name: "consumer refers a valid secret",
+			modifyBasicConsumer: func(c *configurationv1.KongConsumer) {
+				c.Credentials = []string{"secret"}
+			},
+			secrets:   []*corev1.Secret{validSecret()},
+			operation: admissionv1.Create,
+			expectOK:  true,
+		},
+		{
+			name: "consumer refers a secret with no kongCredType",
+			modifyBasicConsumer: func(c *configurationv1.KongConsumer) {
+				c.Credentials = []string{"secret"}
+			},
+			secrets: []*corev1.Secret{
+				func() *corev1.Secret {
+					s := validSecret()
+					delete(s.Data, "kongCredType")
+					return s
+				}(),
+			},
+			operation:       admissionv1.Create,
+			expectError:     false,
+			expectedMessage: "consumer credential failed validation: missing required key kongCredType",
+		},
+		{
+			name: "consumer refers a secret that is already referred by another consumer",
+			modifyBasicConsumer: func(c *configurationv1.KongConsumer) {
+				c.Credentials = []string{"secret"}
+			},
+			secrets: []*corev1.Secret{validSecret()},
+			consumers: []*configurationv1.KongConsumer{
+				func() *configurationv1.KongConsumer {
+					c := basicConsumer()
+					c.Name = "consumer-2"
+					c.Credentials = []string{"secret"}
+					return &c
+				}(),
+			},
+			operation:       admissionv1.Create,
+			expectError:     false,
+			expectedMessage: "consumer credential violated unique key constraint: unique key constraint violated for key",
+		},
+		{
+			name: "consumer refers a secret that has the same key as another secret",
+			modifyBasicConsumer: func(c *configurationv1.KongConsumer) {
+				c.Credentials = []string{"secret"}
+			},
+			secrets: []*corev1.Secret{
+				validSecret(),
+				func() *corev1.Secret {
+					s := validSecret()
+					s.Name = "secret-2"
+					return s
+				}(),
+			},
+			consumers: []*configurationv1.KongConsumer{
+				func() *configurationv1.KongConsumer {
+					c := basicConsumer()
+					c.Name = "consumer-2"
+					c.Credentials = []string{"secret-2"}
+					return &c
+				}(),
+			},
+			operation:       admissionv1.Create,
+			expectError:     false,
+			expectedMessage: "consumer credential violated unique key constraint: unique key constraint violated for key",
```

Comment:
A bit of a nitpick but this isn't ideal. The enclosed message is duplicated and doesn't read well. 🤔

Reply:
9451732